### PR TITLE
enable auto-completion

### LINKED ISSUE #426: repl: autocomplete
using https://codemirror.net/examples/autocompletion/ the editor should have an autocomplete that uses the data from jsdoc.

### PR DESCRIPTION
**Pull request by roipoussiere**
Now https://github.com/uiwjs/react-codemirror/pull/458 is merged, the auto-completion seems to work pretty well, so maybe we can enable it?
follows #427, fix #426

### CODE REVIEW

**Comment by roipoussiere** [comment]
update: css fix to avoid this glitch:

image

**Comment by yaxu** [comment]
It would be good to have an option for this, personally I'd prefer it default off.

**Comment by felixroos** [comment]
thanks for reviving the autocomplete :) besides the now merged PR (https://github.com/uiwjs/react-codemirror/pull/458), there is one more thing still not done, as mentioned in https://github.com/tidalcycles/strudel/pull/427  :

> todo: be context aware, show methods only after a dot, show top level functions only on top level etc..
see https://codemirror.net/examples/autocompletion/ (Completing from Syntax)

without this, the autocomplete is pretty dumb, as it suggests any name at any place, so many things that get suggested don't actually work, for example:

```js
s("bd sd").s
```

this will suggest `samples`, which is not a pattern method, so it will throw an error. (I haven't tested this specific case but I remember that it did things like that).

maybe it makes sense to add this before merging? it could be a bit tricky to decide how to decide which goes what, and how to annotate it.
but we could also add it as an opt-in experimental feature as it is now..

**Comment by roipoussiere** [comment]
I'm adding the button in the settings, so I can disable the feature by default until he problem you mention is not solved.

>  but we could also add it as an opt-in experimental feature as it is now..

**Comment by roipoussiere** [comment]
> It would be good to have an option for this, personally I'd prefer it default off.

Done

File: packages/react/src/components/CodeMirror6.jsx [hunk 1]
```diff
@@ -140,6 +142,12 @@ export default function CodeMirror({
     }
     return staticExtensions;
   }, [keybindings]);
+
+  const setAutoCompletion = (isAutoCompletionEnabled) =>
```
Comment (felixroos) [inline packages/react/src/components/CodeMirror6.jsx:146]: why is this a function? couldn't this be added to extensions inside the useMemo (L135+) ?
outside of the memo the function will be created and the extensions will be spreaded on every render.
This creates a little bit of garbage on every keystroke

Comment (roipoussiere) [reply packages/react/src/components/CodeMirror6.jsx]: Indeed, fixed

**Comment by felixroos** [comment]
I'll merge this now with the current state, although it's currently sometimes suggesting things that are wrong, it's not on by default and already has value as-is.

**Pull request merged**